trading_calendars/exchange_calendar_xpsx.py

```diff
@@ -0,0 +1,394 @@
+#
+# Copyright 2019 Quantopian, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import time
+from itertools import chain
+
+import pandas as pd
+from pandas.tseries.holiday import Holiday
+from pytz import timezone
+
+from .common_holidays import european_labour_day
+from .trading_calendar import HolidayCalendar, TradingCalendar
+
+
+class XPSXExchangeCalendar(TradingCalendar):
+    """
+    Exchange calendar for the Pakistan Stock Exchange (XPSX).
+
+    Open Time:
+    Close Time:
+
+    Regularly-Observed Holidays:
+    - Kashmir Day (Feb 5)
+    - Pakistan Day (Mar 23)
+    - Labour Day (May 1)
+    - Juma-Tul-Wida (last Friday of Ramadan)
+    - Eil-ul-Fitr (1st-3rd Shawwal)
+    - Eid-ul-Azha (10th-11th Zil-Hajj)
+    - Independence Day (Aug 14)
+    - Ashura (9th & 10th Muharram)
+    - Eid Milad-un-Nabi (12th Rabi-ul-Awal)
+    - Birthday of Quaid-e-Azam & Christmas (Dec 25)
+
+    Formerly-Observed Holidays:
+    - Iqbal Day (Nov 9; ended in 2012)
+
+    Occasional election and bank holidays are also observed.
+    """
+    name = 'XPSX'
+
+    tz = timezone('Asia/Karachi')
+
+    open_times = (
+        (None, time(9, 32)),
+        # TODO: 09:17 on Fridays
+    )
+
+    close_times = (
+        (None, time(15, 30)),
+        # TODO: 16:30 on Fridays
+    )
+
+    @property
+    def regular_holidays(self):
+        return HolidayCalendar([
+            Holiday(
+                "Kashmir Day",
+                month=2,
+                day=5,
+            ),
+            Holiday(
+                "Pakistan Day",
+                month=3,
+                day=23,
+            ),
+            european_labour_day(),
+            Holiday(
+                "Independence Day",
+                month=8,
+                day=14,
+            ),
+            Holiday(
+                "Iqbal Day",
+                month=11,
+                day=9,
+                end_date='2013',
+            ),
+            Holiday(
+                "Birthday of Quaid-e-Azam & Christmas",
+                month=12,
+                day=25,
+            ),
+        ])
+
+    juma_tul_wida = pd.to_datetime([
+        # Juma-Tul-Wida (last Friday of Ramadan before Eid-ul-Fitr)
+        '2004-11-11',  # NOTE: Observed on Thursday instead of Friday (??)
+        # Not observed in 2009 or 2010.
```

Comment:
What about the other years between 2004 and 2013?

Reply:
Good catch: looks like this is actually a *mostly* annual observance, but some of the other years had been mislabeled as Eid-ul-Fitr, or were unreported and found via analyzing the trading volume. Still not sure what happened in 2009 :\